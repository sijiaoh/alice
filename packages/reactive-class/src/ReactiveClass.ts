export interface Store {
  onChange: () => void;
}

export interface ReactiveClassConstructorParams {
  store: Store;
}

export class ReactiveClass {
  constructor({ store }: ReactiveClassConstructorParams) {
    this.store = store;
    return new Proxy(this, {
      set: (target, p: keyof ReactiveClass, value: never) => {
        // eslint-disable-next-line no-param-reassign
        target[p] = value;
        this.store.onChange();
        return true;
      },
    });
  }

  private store: Store;
}
