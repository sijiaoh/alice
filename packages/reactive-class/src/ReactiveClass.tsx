/* eslint-disable react-hooks/rules-of-hooks */

import autobind from 'autobind-decorator';
import { produce } from 'immer';
import React, { useEffect, useState } from 'react';
import { Props } from 'react-utils';
import {
  atom,
  RecoilState,
  useRecoilCallback,
  RecoilValue,
  Loadable,
  useRecoilState,
  RecoilRoot,
  useRecoilTransactionObserver_UNSTABLE,
} from 'recoil';

@autobind
export class ReactiveClass<T> {
  static readonly notProvideError = 'ReactiveClass.Provider not provided.';
  static get?: <U>(recoilValue: RecoilValue<U>) => Loadable<U>;
  static set?: <U>(
    recoilVal: RecoilState<U>,
    valOrUpdater: U | ((currVal: U) => U)
  ) => void;

  static Component({ children }: Props) {
    const dummy = useRecoilCallback(({ snapshot, set }) => {
      ReactiveClass.get = snapshot.getLoadable;
      ReactiveClass.set = set;
      return () => {};
    });

    useRecoilTransactionObserver_UNSTABLE(({ snapshot }) => {
      ReactiveClass.get = snapshot.getLoadable;
    });

    const [initialized, setInitialized] = useState(false);
    useEffect(() => {
      // jest用呼び出し。
      dummy();
      setInitialized(true);
    }, [dummy]);

    return <>{initialized ? children : null}</>;
  }

  static Provider({ children }: Props) {
    const { Component } = ReactiveClass;
    return (
      <RecoilRoot>
        <Component>{children}</Component>
      </RecoilRoot>
    );
  }

  readonly state: RecoilState<T>;

  get data() {
    return this.getData();
  }
  set data(data: Readonly<T>) {
    this.setData(data);
  }

  constructor(key: string, defaultStore: Readonly<T>) {
    this.state = atom({ key, default: defaultStore });
  }

  getData() {
    if (!ReactiveClass.get) throw new Error(ReactiveClass.notProvideError);
    return ReactiveClass.get(this.state).getValue();
  }

  setData(state: Readonly<T>) {
    if (!ReactiveClass.set) throw new Error(ReactiveClass.notProvideError);
    ReactiveClass.set(this.state, state);
  }

  changeValue(callback: (state: T) => Readonly<T> | void) {
    this.setData(produce(this.getData(), callback));
  }

  useState() {
    return useRecoilState(this.state);
  }
}
