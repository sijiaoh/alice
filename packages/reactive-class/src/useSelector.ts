import { useState, useEffect } from 'react';
import { ReactiveClass } from './ReactiveClass';

export const useSelector = <T extends ReactiveClass, R>(
  v: T,
  callback: (v: T) => R
) => {
  const [value, setValue] = useState(callback(v));

  useEffect(() => {
    return v.subscribe((lv) => {
      setValue(callback(lv));
    });
  }, []);

  return value;
};
