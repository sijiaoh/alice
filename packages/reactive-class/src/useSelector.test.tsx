import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Subject } from 'rxjs';
import { ReactiveClass } from './ReactiveClass';
import { useSelector } from './useSelector';

describe('useSelector', () => {
  it('will update render', () => {
    class Some extends ReactiveClass {
      prop = 'a';
    }
    const some = new Some();

    const Component = ({ s }: { s: Some }) => {
      const prop = useSelector(s, (ls) => ls.prop);
      return <>{prop}</>;
    };

    let component!: ReactTestRenderer;
    void act(() => {
      component = create(<Component s={some} />);
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some as any).subject as Subject<Some>).observers.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();

    void act(() => {
      some.prop = 'b';
    });
    expect(component.toJSON()).toMatchSnapshot();

    void act(() => {
      component.unmount();
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect(((some as any).subject as Subject<Some>).observers.length).toBe(0);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
