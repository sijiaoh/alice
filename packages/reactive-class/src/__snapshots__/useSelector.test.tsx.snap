// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSelector will update render 1`] = `"a"`;

exports[`useSelector will update render 2`] = `"b"`;

exports[`useSelector will update render 3`] = `null`;
